File: JavaScript/207-Course Schedule.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {boolean}
 */
var canFinish = function(numCourses, prerequisites) {
    let indegree = new Array(numCourses).fill(0);    
    let graph = new Map();
    
    for (let [a, b] of prerequisites){
        graph.set(b, (graph.get(b) || []).concat([a]));
        indegree[a] += 1;
    }
    
    let queue = new Array();
    for (let i=0; i<numCourses; i++){
        if (indegree[i] === 0) queue.push(i);
    }
    
    while (queue.length){
        let size = queue.length;
        // console.log(queue)
        for (let j=0; j<queue.length; j++){
            let curr = queue.shift();
            if (graph.get(curr)){
                for (let nxt of graph.get(curr)){
                    indegree[nxt] -= 1;
                    if (indegree[nxt] === 0){
                        queue.push(nxt);
                    }
                }
                graph.delete(curr);
            }
        }
    }
    
    if (graph.size === 0){
        return true;
    }else{
        return false;
    }
};

File: C++/207-Course Schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> graph(numCourses);
        vector<int> indegree(numCourses);
        for (auto p:prerequisites) {
            int u = p[0], v = p[1];
            graph[v].push_back(u);
            indegree[u]++;
        }

        queue<int> q;

        for (int i=0; i<numCourses; i++) {
            if (indegree[i] == 0) q.push(i);
        }

        int cnt = 0;
        while (!q.empty()) {
            int curr = q.front();
            q.pop();
            cnt++;
            for (int nxt:graph[curr]) {
                indegree[nxt]--;
                if (indegree[nxt] == 0) q.push(nxt);
            }
        }

        return cnt == numCourses;
    }
};